<a href="https://colab.research.google.com/github/asis012/bert/blob/master/QNLI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import sys, os
from pathlib import Path
if not "cwd" in globals():
    cwd = Path(os.getcwd())
cwd



PosixPath('/root')

In [0]:
!pip install git+git://github.com/HazyResearch/snorkel.git@snorkel-superglue

  Cloning git://github.com/HazyResearch/snorkel.git (to revision snorkel-superglue) to /tmp/pip-req-build-h47je2te
  Running command git clone -q git://github.com/HazyResearch/snorkel.git /tmp/pip-req-build-h47je2te
  Running command git checkout -q eb9ba3476eeb932f31ed92e5cc7dfa475611216e
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 10.1MB 2.7MB/s 
     |████████████████████████████████| 17.3MB 12.7MB/s 
     |████████████████████████████████| 153kB 41.2MB/s 
     |████████████████████████████████| 1.7MB 28.7MB/s 
     |████████████████████████████████| 12.9MB 28.5MB/s 
     |████████████████████████████████| 716kB 39.8MB/s 
     |████████████████████████████████| 12.5MB 32.7MB/s 
     |████████████████████████████████| 26.6MB 1.5MB/s 
     |████████████████████████████████| 14.3MB 31.8MB/s 
     |████████████████████████████████| 133kB 40.6MB/s 
     |█████████████

In [0]:
cd tmp

/tmp


In [0]:
ls

In [0]:
pwd

'/tmp'

In [0]:
cd ..

/


In [0]:
ls

sample_data/


In [0]:
cd ..

/


In [0]:
ls

bin/      dev/   lib32/  opt/   sbin/   tensorflow-2.0.0b1/  var/
boot/     etc/   lib64/  proc/  srv/    tmp/
content/  home/  media/  root/  swift/  tools/
datalab/  lib/   mnt/    run/   sys/    usr/


In [0]:
cd root

/root


In [0]:
cd ..

/


In [0]:
cd usr

/usr


In [0]:
cd ..

/


In [0]:
ls

bin/  games/  grte/  include/  lib/  lib32/  local/  sbin/  share/  src/


In [0]:
pwd

'/root'

In [0]:
from snorkel.mtl.data import MultitaskDataLoader
from snorkel.mtl.model import MultitaskModel
from snorkel.mtl.snorkel_config import default_config as config
from snorkel.mtl.trainer import Trainer

ModuleNotFoundError: ignored

In [0]:
import superglue_tasks
from tokenizer import get_tokenizer
from utils import task_dataset_to_dataframe


In [0]:
from dataloaders import get_jsonl_path
from superglue_parsers.wic import get_rows

In [0]:
!unzip WiC.zip

unzip:  cannot find or open WiC.zip, WiC.zip.zip or WiC.zip.ZIP.


In [0]:
ls

cb.py    __init__.py  __pycache__/  swag.py  wsc.py
copa.py  multirc.py   rte.py        wic.py


In [0]:
task_name= "WiC"
data_dir = os.environ.get("SUPERGLUEDATA", os.path.join(str(cwd.parents[0]), "data"))
split = "valid"



In [0]:
data_df['sentence1_idx'] = ""
  
# Observe the result 


In [0]:
data_df['sentence2_idx'] = ""


In [0]:
data_df

end1  end2  idx  ...               word sentence1_idx sentence2_idx
0    30    90    0    ...  class                                        
1    27    25    1    ...  stripe                                       
2    36    7     2    ...  check                                        
3    31    35    3    ...  brush                                        
4    23    52    4    ...  acquisition                                  
5    7     6     5    ...  thing                                        
6    12    38    6    ...  minister                                     
7    17    21    7    ...  easiness                                     
8    30    48    8    ...  contradistinction                            
9    79    8     9    ...  home                                         
10   24    14    10   ...  inclination                                  
11   12    13    11   ...  assurance                                    
12   40    32    12   ...  save                                         
13   13    21    13   ...  bang                                         
14   17    23    14   ...  tremor                                       
15   36    13    15   ...  response                                     
16   18    48    16   ...  touch                                        
17   4     13    17   ...  hail                                         
18   26    19    18   ...  rail                                         
19   34    40    19   ...  obstacle                                     
20   19    32    20   ...  socialization                                
21   42    22    21   ...  rank                                         
22   23    4     22   ...  bend                                         
23   18    14    23   ...  buy                                          
24   5     9     24   ...  youth                                        
25   6     3     25   ...  lay                                          
26   5     11    26   ...  leave                                        
27   22    30    27   ...  aliyah                                       
28   7     22    28   ...  violate                                      
29   17    39    29   ...  pencil                                       
..   ..    ..    ..   ...     ...            ..            ..           
608  15    5     608  ...  pound                                        
609  6     9     609  ...  donate                                       
610  3     59    610  ...  hew                                          
611  29    7     611  ...  breed                                        
612  14    8     612  ...  risk                                         
613  30    41    613  ...  sit                                          
614  12    60    614  ...  review                                       
615  19    13    615  ...  jockey                                       
616  14    16    616  ...  body                                         
617  26    22    617  ...  stabilize                                    
618  11    64    618  ...  lead                                         
619  48    5     619  ...  base                                         
620  45    14    620  ...  inadequacy                                   
621  38    43    621  ...  upgrade                                      
622  37    49    622  ...  criticism                                    
623  16    9     623  ...  bastion                                      
624  11    39    624  ...  pity                                         
625  12    34    625  ...  embellish                                    
626  16    31    626  ...  tilt                                         
627  42    22    627  ...  facility                                     
628  15    24    628  ...  hour                                         
629  16    17    629  ...  capital                                      
630  25    16    630  ...  fire                                

In [0]:
task_name = "WiC"
data_dir = os.environ.get("SUPERGLUEDATA", os.path.join(str(cwd.parents[0]), "./snorkel-superglue-master/Data"))
split = "valid"
max_data_samples = None # max examples to include in dataset

jsonl_path = get_jsonl_path(data_dir, task_name, split)
# print(jsonl_path)
# data_df = pd.read_json(jsonl_path, lines=True)

wic_df = pd.DataFrame.from_records(get_rowss(jsonl_path, max_data_samples))

wic_df[["sentence1_idx"]].head()



NameError: ignored

In [0]:
import json


from pytorch_pretrained_bert import BertTokenizer


import superglue_parsers

In [0]:
def get_rowss(jsonl_path, max_data_samples):
  
  logger.info(f"Loading data from {jsonl_path}.")
  
  rows = [json.loads(row) for row in open(jsonl_path, encoding="utf-8")]
  
      
  print(rows)    
      
  for i in range(2):
    logger.info(f"Sample {i}: {rows[i]}")
#     print(open(jsonl_path, encoding="utf-8"))
    # Truncate to max_data_samples
  
  
  if max_data_samples:
    rows = rows[:max_data_samples]
    logger.info(f"Truncating to {max_data_samples} samples.")

  for row in rows:
    row["sentence1_idx"] = int(row["idx"])
    row["sentence2_idx"] = int(row["idx"])
    row["label"] = row["label"] if "label" in row else True
    
  return rows
import logging  
logger = logging.getLogger(__name__)

def get_ro(jsonl_path, max_data_samples):
    logger.info(f"Loading data from {jsonl_path}.")
    rows = [json.loads(row) for row in open(jsonl_path, encoding="utf-8")]
    for i in range(2):
        logger.info(f"Sample {i}: {rows[i]}")
    print(rows)
    # Truncate to max_data_samples
    if max_data_samples:
        rows = rows[:max_data_samples]
        logger.info(f"Truncating to {max_data_samples} samples.")

    for row in rows:
        row["sentence1_idx"] = int(row["idx"])
        row["sentence2_idx"] = int(row["idx"])
        row["label"] = row["label"] if "label" in row else True

    return rows


In [0]:
bert_model = "bert-large-cased"
tokenizer_name = "bert-large-cased"
batch_size = 4
max_sequence_length = 256

In [0]:
tokenizer = get_tokenizer(tokenizer_name)
print(max_data_samples)

None


In [0]:

from dataloaders import get_dataloaders

dataloaders = get_dataloaders(
    task_name=TASK_NAME,
    tokenizer_name=BERT_MODEL,
    **dataloader_config
)

In [0]:
def get_dataseet(
    data_dir: str,
    task_name: str,
    split: str,
    tokenizer: BertTokenizer,
    max_data_samples: int,
    max_sequence_length: int,
):
    jsonl_path = get_jsonl_path(data_dir, task_name, split)
    return superglue_parsers.parser[task_name](
        jsonl_path, tokenizer, max_data_samples, max_sequence_length
    )


In [0]:
base_task = superglue_tasks.task_funcs[task_name](bert_model)
tasks = [base_task]
tasks

In [0]:
!rm wic.py

In [0]:
!unzip RTE.zip

Archive:  RTE.zip
   creating: RTE/
  inflating: RTE/train.jsonl         
   creating: __MACOSX/RTE/
  inflating: __MACOSX/RTE/._train.jsonl  
  inflating: RTE/test.jsonl          
  inflating: __MACOSX/RTE/._test.jsonl  
  inflating: RTE/val.jsonl           
  inflating: __MACOSX/RTE/._val.jsonl  
  inflating: __MACOSX/._RTE          


In [0]:
from google.colab import files
uploaded = files.upload()

MessageError: ignored

When you execute your script, CWD is set to the directory where your script is.

In [0]:
import os, sys
from pathlib import Path

if not "cwd" in globals():
    cwd = Path(os.getcwd())


Usine Bert pre trained model.
Out of four model using bert_large-cases


Use ADAM optimer defined by bert

In [0]:

TASK_NAME = "RTE"
BERT_MODEL = "bert-large-cased"

dataloader_config = {
    "batch_size": 8,
    "data_dir": os.environ.get("SUPERGLUEDATA", os.path.join(str(cwd.parents[0]), "./snorkel-superglue-master/Data")),
    "splits": ["train", "valid"],
}

trainer_config = {
    "lr": 2e-5,
    "optimizer": "adamax",
    "n_epochs": 15,
    "conter_unit": "epochs",
    "evaluation_freq": 0.25,
}

load data and return dataloaders

In [0]:

from dataloaders import get_dataloaders

dataloaders = get_dataloaders(
    task_name=TASK_NAME,
    tokenizer_name=BERT_MODEL,
    **dataloader_config
)

dataloaders

In [0]:
from superglue_tasks import task_funcs

task = task_funcs[TASK_NAME](BERT_MODEL)

100%|██████████| 1242874899/1242874899 [00:32<00:00, 38291927.81B/s]


In [0]:
from snorkel.mtl.model import MultitaskModel
from snorkel.mtl.trainer import Trainer


ModuleNotFoundError: ignored

get the pre trained bert  model which will save our time and memory
  if you want to train the bert model:
  use following code
  trainer = Trainer(**trainer_config) 

  trainer.train_model(model, dataloaders)

In [0]:
! wget -nc https://www.dropbox.com/s/t2ri9o0iz765hsn/RTE_bert.pth?dl=0 && mv RTE_bert.pth?dl=0 RTE_bert.pth
model.load("RTE_bert.pth")

--2019-08-09 09:10:16--  https://www.dropbox.com/s/t2ri9o0iz765hsn/RTE_bert.pth?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.1.1, 2620:100:601b:1::a27d:801
Connecting to www.dropbox.com (www.dropbox.com)|162.125.1.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/t2ri9o0iz765hsn/RTE_bert.pth [following]
--2019-08-09 09:10:16--  https://www.dropbox.com/s/raw/t2ri9o0iz765hsn/RTE_bert.pth
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc339f237333998a5c22478adb5c.dl.dropboxusercontent.com/cd/0/inline/AmR5qHOaSgSa1Tt8FAks6hS51jk4ikNyAkPztXiGzIQdOUSxGFy4hdQONfKuHlHmzqSgkm8E5FsxUKV1Gup6KzY-HR5BfyzNdCCVie2Zqzh5zw/file# [following]
--2019-08-09 09:10:16--  https://uc339f237333998a5c22478adb5c.dl.dropboxusercontent.com/cd/0/inline/AmR5qHOaSgSa1Tt8FAks6hS51jk4ikNyAkPztXiGzIQdOUSxGFy4hdQONfKuHlHmzqSgkm8E5FsxUKV1Gup6KzY-HR5BfyzNdCCVie2Zqzh5zw/file
Resolving uc33

NameError: ignored

In [0]:
dataloaders[1]

In [0]:
%%time
model.score(dataloaders[1])

CPU times: user 7min 38s, sys: 7.87 s, total: 7min 46s
Wall time: 7min 47s


{'RTE/SuperGLUE/valid/accuracy': 0.7364620938628159}

Here we have found the accuracy of bert model. which is 0.736.

Now lets use some slicing functions

In [0]:
from snorkel.slicing.sf import slicing_function

In [0]:
@slicing_function()
def slice_temporal_preposition(example):
    temporal_prepositions = ["after", "before", "past"]
    both_sentences = example.sentence1 + example.sentence2
    return any([p in both_sentences for p in temporal_prepositions])

@slicing_function()
def slice_possessive_preposition(example):
    possessive_prepositions = ["inside of", "with", "within"]
    both_sentences = example.sentence1 + example.sentence2
    return any([p in both_sentences for p in possessive_prepositions])


In [0]:
!pip install pytorch_pretrained_bert

!pip install allennlp

!pip install jsonlines

!pip install git+git://github.com/HazyResearch/snorkel.git@snorkel-superglue

!pip install git+https://github.com/HazyResearch/snorkel


  Cloning git://github.com/HazyResearch/snorkel.git (to revision snorkel-superglue) to /tmp/pip-req-build-5449sznl
  Running command git clone -q git://github.com/HazyResearch/snorkel.git /tmp/pip-req-build-5449sznl
  Running command git checkout -q eb9ba3476eeb932f31ed92e5cc7dfa475611216e
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 51kB 3.5MB/s 
     |████████████████████████████████| 12.5MB 4.5MB/s 
     |████████████████████████████████| 5.4MB 33.8MB/s 
     |████████████████████████████████| 10.1MB 5.8MB/s 
     |████████████████████████████████| 14.3MB 16.3MB/s 
     |████████████████████████████████| 12.9MB 35.4MB/s 
     |████████████████████████████████| 716kB 43.4MB/s 
     |████████████████████████████████| 26.6MB 54.1MB/s 
     |████████████████████████████████| 27.7MB 41.9MB/s 
     |████████████████████████████████| 153kB 46.2MB/s 
     |███████████████

  Cloning https://github.com/HazyResearch/snorkel to /tmp/pip-req-build-z8g7rx9r
  Running command git clone -q https://github.com/HazyResearch/snorkel /tmp/pip-req-build-z8g7rx9r
  Created wheel for snorkel: filename=snorkel-0.7.0b0-cp36-none-any.whl size=168807 sha256=1347ef84ae5b89911d8f9f0d075ac0ae97accb9e6478e49cab41ca30d584987f
  Stored in directory: /tmp/pip-ephem-wheel-cache-2ruv3dcw/wheels/62/6f/22/444271063e7be3c3547fd48ba1782c8053489be0e0226990b0
Successfully built snorkel
  Found existing installation: snorkel 0.9.0
    Uninstalling snorkel-0.9.0:
      Successfully uninstalled snorkel-0.9.0


In [0]:
print("hello")

hello


In [0]:
%time

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 5.01 µs


In [0]:


!pip install git+git://github.com/HazyResearch/snorkel.git@snorkel-superglue


  


  Cloning git://github.com/HazyResearch/snorkel.git (to revision snorkel-superglue) to /tmp/pip-req-build-f50m9amm
  Running command git clone -q git://github.com/HazyResearch/snorkel.git /tmp/pip-req-build-f50m9amm
  Running command git checkout -q eb9ba3476eeb932f31ed92e5cc7dfa475611216e
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
  Created wheel for snorkel: filename=snorkel-0.9.0-cp36-none-any.whl size=109793 sha256=133f43a2acfa313c35e277b919e68e31f6f44d9d353fe58ede8986382f5656f7
  Stored in directory: /tmp/pip-ephem-wheel-cache-bdhjme99/wheels/1d/74/e9/0c64265729943b0dab0a757f726491151c4ca9ccea68db8e1a
Successfully built snorkel
  Found existing installation: snorkel 0.9.1+dev
    Uninstalling snorkel-0.9.1+dev:
      Successfully uninstalled snorkel-0.9.1+dev


In [0]:
from snorkel.mtl.model import MultitaskModel
from snorkel.mtl.trainer import Trainer

In [0]:
!pip install snorkel-metal


In [0]:
!unzip snorkel-superglue-master.zip

Archive:  snorkel-superglue-master.zip
   creating: snorkel-superglue-master/
  inflating: snorkel-superglue-master/run.py  
   creating: __MACOSX/
   creating: __MACOSX/snorkel-superglue-master/
  inflating: __MACOSX/snorkel-superglue-master/._run.py  
  inflating: snorkel-superglue-master/.DS_Store  
  inflating: __MACOSX/snorkel-superglue-master/._.DS_Store  
  inflating: snorkel-superglue-master/LICENSE  
  inflating: __MACOSX/snorkel-superglue-master/._LICENSE  
  inflating: snorkel-superglue-master/requirements.txt  
  inflating: __MACOSX/snorkel-superglue-master/._requirements.txt  
  inflating: snorkel-superglue-master/dataloaders.py  
  inflating: __MACOSX/snorkel-superglue-master/._dataloaders.py  
  inflating: snorkel-superglue-master/task_config.py  
  inflating: __MACOSX/snorkel-superglue-master/._task_config.py  
  inflating: snorkel-superglue-master/download_superglue_data.sh  
  inflating: __MACOSX/snorkel-superglue-master/._download_superglue_data.sh  
   creating: sno

In [0]:
import os, sys
import pandas as pd
from pathlib import Path
import json
if not "cwd" in globals():
    cwd = Path(os.getcwd())


In [0]:
TASK_NAME = "QNLI"
BERT_MODEL = "bert-large-cased"
split = "valid"

dataloader_config = {
    "batch_size": 100,
    "data_dir": "Data",
    "splits": ["train", "valid"],
}

trainer_config = {
    "lr": 0.6,
    "optimizer": "adamax",
    "n_epochs": 5,
    "conter_unit": "epochs",
    "evaluation_freq": 0.5,
}

In [0]:
cd snorkel-superglue-master/t

/snorkel-superglue-master/tutorials


In [0]:
ls

COPA_pretrain_tutorial.ipynb*  Untitled1.ipynb
model_with_qnli.pth            Untitled2.ipynb
QNLI.ipynb                     Untitled.ipynb
RTE_slicing_tutorial.ipynb*    WiC_augmentation_tutorial.ipynb*
snorkel.db                     WiC_slicing_tutorial.ipynb*


In [0]:
from dataloaders import get_jsonl_path
from superglue_parsers.wic import get_rows
max_data_samples =1120

In [0]:
cd ..

/snorkel-superglue-master


In [0]:
from dataloaders import get_jsonl_path
from superglue_parsers.wic import get_rows
max_data_samples =1120



In [0]:
from dataloaders import get_dataloaders

dataloaders = get_dataloaders(
    task_name=TASK_NAME,
    tokenizer_name=BERT_MODEL,
    **dataloader_config
)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [0]:
from superglue_tasks import task_funcs

task = task_funcs[TASK_NAME](BERT_MODEL)

100%|██████████| 1242874899/1242874899 [00:21<00:00, 57819268.78B/s]


In [0]:
from snorkel.mtl.model import MultitaskModel
from snorkel.mtl.trainer import Trainer

model = MultitaskModel(tasks=[task], device=-1, dataparallel=False)

In [0]:
model.load("tutorials/model_with_qnli.pth")

In [0]:
%%time
model.score(dataloaders[1])

CPU times: user 6min 34s, sys: 5.14 s, total: 6min 39s
Wall time: 6min 39s


{'QNLI/SuperGLUE/valid/accuracy': 0.58}